src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.Helpers.cs

```diff
@@ -0,0 +1,861 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers.Binary;
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics.X86;
+using Internal.Runtime.CompilerServices;
+
+namespace System.Text.Unicode
+{
+    internal static partial class Utf8Utility
+    {
+        /// <summary>
+        /// Given a machine-endian DWORD which four bytes of UTF-8 data, interprets the
+        /// first three bytes as a three-byte UTF-8 subsequence and returns the UTF-16 representation.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint ExtractCharFromFirstThreeByteSequence(uint value)
+        {
+            if (BitConverter.IsLittleEndian)
```

Comment:
`ExtractCharFromFirstTwo...` asserts the pattern matched, should this one as well?